Confusing failure when setting up standard streams for Providers.exec and javaexec

### Current Behavior

Build fails with a confusing message:
```
FAILURE: Build failed with an exception.

* Where:
Build file '/private/tmp/tests/1/build.gradle' line: 2

* What went wrong:
A problem occurred evaluating root project '1'.
> Could not create provider for value source ProcessOutputValueSource.
```

### Expected Behavior

The clarifying message `Standard streams cannot be configured for exec output provider` should be visible without specifying `--stacktrace`.

### Context (optional)

_No response_

### Self-contained Reproducer Project

```
providers.exec {
  standardInput = System.in
  commandLine("echo")
}
```

### Gradle version

n/a

### Build scan URL (optional)

_No response_

### Your Environment (optional)

_No response_

--- a/platforms/core-configuration/model-core/src/integTest/groovy/org/gradle/api/provider/ProcessOutputProviderIntegrationTest.groovy
+++ b/platforms/core-configuration/model-core/src/integTest/groovy/org/gradle/api/provider/ProcessOutputProviderIntegrationTest.groovy
@@ -20,6 +20,7 @@ import org.gradle.integtests.fixtures.AbstractIntegrationSpec
 import org.gradle.process.ShellScript
 import org.gradle.process.TestJavaMain
 import org.gradle.util.internal.TextUtil
+import spock.lang.Issue
 
 class ProcessOutputProviderIntegrationTest extends AbstractIntegrationSpec {
     def "providers.exec can be used during configuration time"() {
@@ -262,6 +263,25 @@ class ProcessOutputProviderIntegrationTest extends AbstractIntegrationSpec {
         then:
         outputContains("Other script output")
         result.assertTaskScheduled(":printScriptOutput")
+    }
+
+    @Issue("https://github.com/gradle/gradle/issues/33858")
+    def "providers.exec reports a clear cause when configuring standard streams"() {
+        given:
+        buildFile """
+            providers.exec {
+                standardInput = System.in
+                commandLine("echo")
+            }
+
+            task empty() {}
+        """
+
+        when:
+        fails(":empty")
+
+        then:
+        failure.assertHasCause("Standard streams cannot be configured for exec output provider")
     }
 
     static String cmdToExecConfig(String... args) {

--- a/platforms/core-configuration/model-core/src/main/java/org/gradle/api/internal/provider/sources/process/ProviderCompatibleBaseExecSpec.java
+++ b/platforms/core-configuration/model-core/src/main/java/org/gradle/api/internal/provider/sources/process/ProviderCompatibleBaseExecSpec.java
@@ -16,6 +16,7 @@
 
 package org.gradle.api.internal.provider.sources.process;
 
+import org.gradle.api.InvalidUserCodeException;
 import org.gradle.process.BaseExecSpec;
 import org.gradle.process.ProcessForkOptions;
 import org.jspecify.annotations.Nullable;
@@ -36,32 +37,32 @@ abstract class ProviderCompatibleBaseExecSpec implements DelegatingBaseExecSpec 
 
     @Override
     public BaseExecSpec setStandardInput(InputStream inputStream) {
-        throw new UnsupportedOperationException("Standard streams cannot be configured for exec output provider");
+        throw new InvalidUserCodeException("Standard streams cannot be configured for exec output provider");
     }
 
     @Override
     public InputStream getStandardInput() {
-        throw new UnsupportedOperationException("Standard streams cannot be configured for exec output provider");
+        throw new InvalidUserCodeException("Standard streams cannot be configured for exec output provider");
     }
 
     @Override
     public BaseExecSpec setStandardOutput(OutputStream outputStream) {
-        throw new UnsupportedOperationException("Standard streams cannot be configured for exec output provider");
+        throw new InvalidUserCodeException("Standard streams cannot be configured for exec output provider");
     }
 
     @Override
     public OutputStream getStandardOutput() {
-        throw new UnsupportedOperationException("Standard streams cannot be configured for exec output provider");
+        throw new InvalidUserCodeException("Standard streams cannot be configured for exec output provider");
     }
 
     @Override
     public BaseExecSpec setErrorOutput(OutputStream outputStream) {
-        throw new UnsupportedOperationException("Standard streams cannot be configured for exec output provider");
+        throw new InvalidUserCodeException("Standard streams cannot be configured for exec output provider");
     }
 
     @Override
     public OutputStream getErrorOutput() {
-        throw new UnsupportedOperationException("Standard streams cannot be configured for exec output provider");
+        throw new InvalidUserCodeException("Standard streams cannot be configured for exec output provider");
     }
 
     @Override

--- a/platforms/core-configuration/model-core/src/test/groovy/org/gradle/api/internal/provider/sources/process/ProviderCompatibleBaseExecSpecTestBase.groovy
+++ b/platforms/core-configuration/model-core/src/test/groovy/org/gradle/api/internal/provider/sources/process/ProviderCompatibleBaseExecSpecTestBase.groovy
@@ -16,6 +16,7 @@
 
 package org.gradle.api.internal.provider.sources.process
 
+import org.gradle.api.InvalidUserCodeException
 import org.gradle.process.BaseExecSpec
 import org.gradle.test.fixtures.file.TestNameTestDirectoryProvider
 import org.gradle.util.TestUtil
@@ -131,21 +132,48 @@ abstract class ProviderCompatibleBaseExecSpecTestBase extends Specification {
         when:
         specUnderTest.setStandardInput(new ByteArrayInputStream())
         then:
-        thrown UnsupportedOperationException
+        def e = thrown(InvalidUserCodeException)
+        e.message == "Standard streams cannot be configured for exec output provider"
     }
 
     def "setting output stream is forbidden"() {
         when:
         specUnderTest.setStandardOutput(new ByteArrayOutputStream())
         then:
-        thrown UnsupportedOperationException
+        def e = thrown(InvalidUserCodeException)
+        e.message == "Standard streams cannot be configured for exec output provider"
     }
 
     def "setting error stream is forbidden"() {
         when:
         specUnderTest.setErrorOutput(new ByteArrayOutputStream())
         then:
-        thrown UnsupportedOperationException
+        def e = thrown(InvalidUserCodeException)
+        e.message == "Standard streams cannot be configured for exec output provider"
+    }
+
+    def "getting input stream is forbidden"() {
+        when:
+        specUnderTest.getStandardInput()
+        then:
+        def e = thrown(InvalidUserCodeException)
+        e.message == "Standard streams cannot be configured for exec output provider"
+    }
+
+    def "getting output stream is forbidden"() {
+        when:
+        specUnderTest.getStandardOutput()
+        then:
+        def e = thrown(InvalidUserCodeException)
+        e.message == "Standard streams cannot be configured for exec output provider"
+    }
+
+    def "getting error stream is forbidden"() {
+        when:
+        specUnderTest.getErrorOutput()
+        then:
+        def e = thrown(InvalidUserCodeException)
+        e.message == "Standard streams cannot be configured for exec output provider"
     }
 
     def "spec without working directory doesn't set it on parameters"() {

